pkg/dockerregistry/server/blobdescriptorservice.go

```diff
@@ -25,6 +38,155 @@ type blobDescriptorService struct {
 	distribution.BlobDescriptorService
 }
 
+// Stat returns a a blob descriptor if the given blob is either linked in repository or is referenced in
+// corresponding image stream. This method is invoked from inside of upstream's linkedBlobStore. It expects
+// a proper repository object to be set on given context by upper openshift middleware wrappers.
 func (bs *blobDescriptorService) Stat(ctx context.Context, dgst digest.Digest) (distribution.Descriptor, error) {
-	return dockerRegistry.BlobStatter().Stat(ctx, dgst)
+	repo, repoFound := RepositoryFrom(ctx)
+	if !repoFound {
+		context.GetLogger(ctx).Errorf("failed to retrieve repository from context")
+	}
+
+	// if there is a repo layer link, return its descriptor
+	desc, err := bs.BlobDescriptorService.Stat(ctx, dgst)
+	if err == nil {
+		if repoFound {
+			// and remember the association
+			repo.cachedLayers.RememberDigest(dgst, repo.blobrepottl, imageapi.DockerImageReference{
+				Namespace: repo.namespace,
+				Name:      repo.name,
+			}.Exact())
+		}
+		return desc, err
+	}
+
+	if !repoFound {
+		return distribution.Descriptor{}, distribution.ErrBlobUnknown
+	}
+
+	// verify the blob is stored locally
+	desc, err = dockerRegistry.BlobStatter().Stat(ctx, dgst)
+	if err != nil {
+		return desc, err
+	}
+
+	if imageStreamHasBlob(repo, dgst) {
+		return desc, nil
+	}
+
+	return distribution.Descriptor{}, distribution.ErrBlobUnknown
+}
+
+func (bs *blobDescriptorService) Clear(ctx context.Context, dgst digest.Digest) error {
+	if repo, found := RepositoryFrom(ctx); found {
+		repo.cachedLayers.ForgetDigest(dgst, imageapi.DockerImageReference{
+			Namespace: repo.namespace,
+			Name:      repo.name,
+		}.Exact())
+	} else {
+		context.GetLogger(ctx).Errorf("failed to retrieve repository from context")
```

Comment:
> is this really an error?

If it's an error in `Stat`, it's certainly an error here. Our `errorBlobStore` must make sure that repository is set for each and every method call. If unset, there's an error in our code.

> and if so, should we still be calling bs.BlobDescriptorService.Clear(ctx, dgst)? what does clearing the descriptor do? would that affect other users of this blob?

I'm not sure. In this case we clear layer link pointing to a blob. That means the blob won't be gettable from the repository unless it stays referenced in corresponding image stream.

Since I'll be reworking the Stat call to fail immediately on repository unset, I'll do the same here.